core/src/OpenACC/Kokkos_OpenACC_Atomic.hpp

```diff
@@ -0,0 +1,1964 @@
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+__inline__  unsigned int atomic_compare_exchange(
+    volatile unsigned int *const dest, const unsigned int &compare,
+    const unsigned int &val) {
+    unsigned int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicCAS((unsigned int *)dest, compare, val);
+#else
+	unsigned int *ptr = const_cast<unsigned int *>(dest);
+#warning atomic_compare_exchange(unsigned int) not correctly implemented in OpenACC 
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+__inline__  unsigned long long int atomic_compare_exchange(
+    volatile unsigned long long int *const dest, const unsigned long long int &compare,
+    const unsigned long long int &val) {
+    unsigned long long int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicCAS((unsigned long long int *)dest, compare, val);
+#else
+	unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+#warning atomic_compare_exchange(unsigned long long int) not correctly implemented in OpenACC 
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+// atomic_assign ---------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__ void atomic_assign(
+    volatile T *const dest,
+    const T & val) {
+	T *ptr = const_cast<T *>(dest);
+//#pragma acc atomic write
+    ptr[0] = val;
+}
+
+#pragma acc routine seq
+template <>
+ __inline__ void atomic_assign(volatile int *const dest, const int &val) {
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	atomicExch((int *)dest, val);
+#else
+	int *ptr = const_cast<int *>(dest);
+#pragma acc atomic write
+    ptr[0] = val;
+#endif
+}
+
+#pragma acc routine seq
+template <>
+ __inline__ void atomic_assign(volatile unsigned int *const dest, const unsigned int &val) {
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	atomicExch((unsigned int *)dest, val);
+#else
+  unsigned int *ptr = const_cast<unsigned int *>(dest);
+#pragma acc atomic write
+    ptr[0] = val;
+#endif
+}
+
+#pragma acc routine seq
+template <>
+ __inline__ void atomic_assign(volatile unsigned long long int *const dest, const unsigned long long int &val) {
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	atomicExch((unsigned long long int *)dest, val);
+#else
+  unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+#pragma acc atomic write
+    ptr[0] = val;
+#endif
+}
+
+// // atomic_fetch_add ------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+inline  T atomic_fetch_add(
+    volatile T *const dest,
+    const T &val) {
+    T retval;
+	T *ptr = const_cast<T *>(dest);
+//#pragma acc atomic capture
```

Comment:
That code was for unsupported atomic data types; I changed the code so that for unsupported atomic data types, it will abort if running on a device but it will sequentially execute the construct (without atomicity) if running on a host.